.github/workflows/build_and_test.yml

```diff
@@ -15,99 +15,178 @@ concurrency:
   cancel-in-progress: true
 
 jobs:
+  check_code_changes:
+    name: Check Code Changes
+    uses: ./.github/workflows/_check_code_changes.yml
+    with:
+      event_name: ${{ github.event_name }}
+      # For pull_request, use PR's base and head. For push, use event's before and sha.
+      base_sha: ${{ github.event_name == 'pull_request' && github.event.pull_request.base.sha || github.event.before }}
+      head_sha: ${{ github.event_name == 'pull_request' && github.event.pull_request.head.sha || github.sha }}
 
   get-torch-commit:
     runs-on: ubuntu-24.04
+    needs: [check_code_changes]
     outputs:
       torch_commit: ${{ steps.commit.outputs.torch_commit }}
     steps:
-      - id: commit
-        name: Get latest torch commit
+      - name: Get latest torch commit
+        id: commit
+        if: needs.check_code_changes.outputs.has_code_changes == 'true'
         run: |
           echo "torch_commit=$(git ls-remote https://github.com/pytorch/pytorch.git HEAD | awk '{print $1}')" >> "$GITHUB_OUTPUT"
+      - name: Report no code changes
+        if: needs.check_code_changes.outputs.has_code_changes == 'false'
+        run: |
+          echo "No code changes were detected that require running the full test suite."
 
   build-torch-xla:
     name: "Build PyTorch/XLA"
     uses: ./.github/workflows/_build_torch_xla.yml
-    needs: get-torch-commit
-    with:
-      dev-image: us-central1-docker.pkg.dev/tpu-pytorch-releases/docker/development:3.10_tpuvm
-      torch-commit: ${{needs.get-torch-commit.outputs.torch_commit}}
-      timeout-minutes: 240
-    secrets:
-      gcloud-service-key: ${{ secrets.GCLOUD_SERVICE_KEY }}
+    needs: [check_code_changes, get-torch-commit]
+    steps:
```

Comment:
Probably just add an if statement here checking for code changes instead of steps and add a separate common stage for skip build notification instead of adding it to every step. I understand that we might need finer control to skip individual steps but for this PR, this might be sufficient

```
skip-build-notification:
    name: "Skip Build Notification"
    runs-on: ubuntu-latest
    needs: check-code-changes # Depends on the change detection job
    if: needs.check-code-changes.outputs.has_changes != 'true'
    steps:
      - name: Echo no changes message
        run: echo "No code changes detected. Skipping PyTorch/XLA build."
```

Reply:
My understanding is that each of these jobs are required to run. Having the if statement will bypass running them leaving the CI in pending state.

Reply:
yeah, it's a github quirk i think